### DISCUSSION TITLE
Regarding the usage of xtd::forms::busy_box

### DISCUSSION THREAD

**Discussion by baderouaich** [gammasoft71/xtd#243] [Q&A] [answered]
Hello! I hope you are doing fine!

I have a question regarding `xtd::forms::busy_box` usage.

It is necessary to call `xtd::application::do_events()` between `xtd::forms::busy_box::show()` and `xtd::forms::busy_box::hide()` in order for the `busy_box` to be visible, which raises some confusion on where exactly do_events should be called.

for example:
- This one is fine because there are no sub functions and the code is simple:
```cpp
form1() {
    text("Busy box example");
    
    button1.auto_size(true);
    button1.location({10, 10});
    button1.parent(*this);
    button1.text("Do something...");
    button1.click += [] {
      busy_box::show("Please wait while do something...", "Application busy");
      for (auto count = 0; count < 500; ++count) {
        application::do_events();
        thread::sleep(10);
      }
      busy_box::hide();
    };
  }
```

- But this one is bit more confusing for me, for example:
```cpp
form1(){
busy_box::show(*this, "Initializing...", "form1");

  startup_panel_.auto_scroll(true);
  startup_panel_.parent(*this);
  startup_panel_.size(client_size() - drawing::size {0, 100});

application::do_events(); // <-- i must call this here to make busy_box visible
  
  startup_open_recent_projects_list_box_.parent(startup_panel_);
  startup_open_recent_projects_list_box_.location({50, 175});

application::do_events(); // <-- i must call this here to make busy_box visible
  
  some_function();  // <-- should i call application::do_events also from within the function ?

  application::do_events(); // <-- i must call this here to make busy_box visible

  some_function_that_calls_other_init_functions(); // <-- should i call application::do_events() from within all sub-functions ?
  
  busy_box::hide();
}
```

Is there a better approach to use busy_box correctly ? 

Thanks in advance!

**Comment by gammasoft71** [comment, marked as answer]
I'm fine thanks.

First of all the [xtd::application::do_events](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1application.html#a1d4031fa24263d5b25e996ebd2ce9089) method processes all Windows messages currently in the message queue. 
Read the `Remarks` and `Warning`sections for more information about this method.

This is a didactic example, designed to show you how to use the [xtd::forms::message_box](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1busy__box.html) dialog box. 
This example is simple to focused the use of [xtd::forms::busy_box](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1busy__box.html).


The following lines simply emulate a time-consuming process:
```cpp
for (auto count = 0; count < 500; ++count) {
  application::do_events();
  thread::sleep(10);
}
```

But in real code, you'll never use the for loop with the [xtd::threading::thread::sleep](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1threading_1_1thread.html#a42bc6e2bb805f928e50ae73d243b95f2) and [xtd::forms::application::do_event](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1application.html#a1d4031fa24263d5b25e996ebd2ce9089) statements. 
It's a very bad idea to do complex, time-consuming processing in an event.

Instead, create an [xtd::threading::thread](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1threading_1_1thread.html) or, even better, an [xtd::forms::background_worker](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1background__worker.html) in which you can call your business classes. Which can take a lot of time without blocking graphic thread events.

See [How to create a thread ?](https://github.com/gammasoft71/xtd/wiki/How-to-create-a-thread-%3F) wiki for [xtd::threading::thread](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1threading_1_1thread.html) and [xtd::forms::background_worker](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1background__worker.html) usage.

# Examples

The follwing example shows how to use [xtd::forms::busy_box](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1busy__box.html) with [xtd::forms::background_worker](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1background__worker.html) :

```cpp
#include <xtd/forms/application>
#include <xtd/forms/background_worker>
#include <xtd/forms/busy_box>
#include <xtd/forms/button>
#include <xtd/forms/form>
#include <xtd/threading/thread>

using namespace xtd::forms;
using namespace xtd::threading;

class form1 : public form {
public:
  form1() {
    text("Busy box example with background_worker");
    
    backgound_worker1.worker_supports_cancellation(true);
    backgound_worker1.do_work += [] {
      // Call your business class here.
      // the folowing line simulate time consuming process
      thread::sleep(5000);
    };
    
    backgound_worker1.run_worker_completed += [this] {
      busy_box::hide();
      button1.enabled(true);
    };
    
    button1.auto_size(true);
    button1.location({10, 10});
    button1.parent(*this);
    button1.text("Do something...");
    button1.click += [this] {
      button1.enabled(false);
      busy_box::show("Please wait while do something...", "Application busy");
      backgound_worker1.run_worker_async();
    };
    
    form_closed += [this] {
      backgound_worker1.cancel_async();
    };
  }
  
private:
  button button1;
  background_worker backgound_worker1;
};

auto main()->int {
  xtd::forms::application::run(form1 {});
}
```

The follwing example shows how to use [xtd::forms::busy_box](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1busy__box.html) with [xtd::threading::thread](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1threading_1_1thread.html) and [xtd::forms::control::invoke](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1control.html#ae0be25be7221a9405574f064e068d9ef) method :

```cpp
#include <xtd/forms/application>
#include <xtd/forms/busy_box>
#include <xtd/forms/button>
#include <xtd/forms/form>
#include <xtd/threading/thread>

using namespace xtd::forms;
using namespace xtd::threading;

class form1 : public form {
public:
  form1() {
    text("Busy box example with thread");
        
    button1.auto_size(true);
    button1.location({10, 10});
    button1.parent(*this);
    button1.text("Do something...");
    button1.click += [this] {
      button1.enabled(false);
      busy_box::show("Please wait while do something...", "Application busy");
      thread1 = thread {[this] {
        // Call your business class here.
        // the following line simulates a time-consuming process
        thread::sleep(5000);
        
        invoke([this] {
          busy_box::hide();
          button1.enabled(true);
        });
      }};
      thread1.start();
    };
    
    form_closing += [this](object& sender, form_closing_event_args& e) {
      e.cancel(thread1.is_alive());
    };
  }
  
private:
  button button1;
  thread thread1;
};

auto main()->int {
  xtd::forms::application::run(form1 {});
}
```

# Remarks

This explanation applies to all examples using the [xtd::forms::applicaiton::do_events](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1forms_1_1application.html#a1d4031fa24263d5b25e996ebd2ce9089) and [xtd::threading::thread::sleep](https://gammasoft71.github.io/xtd/reference_guides/latest/classxtd_1_1threading_1_1thread.html#a42bc6e2bb805f928e50ae73d243b95f2) methods to simulate a long process.


**I hope that's clearer for you.**

**Reply by baderouaich** [reply]
Thank you so much!! it's clear now :ok_hand:
